src/javascript/reducers/agreementList.js

```diff
@@ -23,135 +25,223 @@ type Action =
+
+  const allAgreementIds: Array<string> = state.allAgreementIds,
+        agreementsById: AgreementListType = state.agreementsById;
+
+  const sortedAgreements: Array<string> = allAgreementIds.sort((a, b) => {
+
+    const aAgreement: string = agreementsById[a][sortKey];
+    const bAgreement: string = agreementsById[b][sortKey];
+
+    if (sortKind === "ascending") {
+      if (aAgreement < bAgreement) return -1;
+      else if (aAgreement > bAgreement) return 1;
+      return 0;
+
+    } else if (sortKind === "descending") {
+      if (aAgreement > bAgreement) return -1;
+      else if (aAgreement < bAgreement) return 1;
+      return 0;
+    };
+    return 0;
+  });
+
+  return sortedAgreements;
+
 };
 
 export default function agreementListReducer(state: AgreementListState = INITIAL_STATE, action: Action): AgreementListState {
   // set agreementId, and check if payload passes an object with an error
-  // or passes just the agreement int. This could be a very long
-
+  // or passes just the agreement id
+  const agreementsById: AgreementListType = state.agreementsById;
   let agreementId: string = '';
+  let agreement: AgreementType;
 
-  // flow really doesn't like this ternary operator, makes me super sad
-  // if (action.payload) {
-  //   agreementId =  typeof action.payload === 'object' ?  action.payload.agreementId.toString(): action.payload.toString();
-  // }
-
-  if (typeof action.payload === 'object') {
+  // REALLY ugly way of making sure that agreementId is in the object, is there
+  // a better solution to this?
+  if (typeof action.payload === 'object' && action.payload.agreementId) {
+    //$FlowFixMe
     agreementId = action.payload.agreementId.toString();
-  } else if (action.payload) {
+  } else if (typeof action.payload === 'string') {
     agreementId = action.payload.toString();
   }
-
-  const agreement: AgreementType = state[agreementId];
+  // once agreementId is there, get the agreementId object
+  // so we can copy with the object spread operator
+  if (agreementId !== '') {
+    agreement = agreementsById[agreementId];
+  }
 
   switch (action.type) {
     case 'REQUEST_PENDING':
       return {
         ...state,
-        [agreementId]: {
-          ...agreement,
-          requestPending: true,
+        agreementsbyId: {
+          ...agreementsById,
+          [agreementId]: {
+            ...agreement,
+            requestPending: true,
+          }
         }
       }
     // send button
     case 'BUYER_SENT':
       return {
         ...state,
-        [agreementId]: {
-          ...agreement,
-          buyerPaid: true,
-          buyerPaidDate: Date.now(),
-          error: "",
-          requestPending: false,
-        },
+        agreementsbyId: {
+          ...agreementsById,
+          [agreementId]: {
+            ...agreement,
+            buyerPaid: true,
+            buyerPaidDate: Date.now(),
```

Comment:
Sweet, I'll look into using web3 to correct this. 

If we were to use the front-end to measure date, would I have to assign `Date.now()` to a variable, then assign `buyerPaidDate`'s value to that variable to keep it constant?